Feature/add client steering flow

CoralMalachi opened the pull request:
According to section 5.8.1 in the EasyMesh Test Plan v1.0 [1], 
add support for passing the Controller Certification
client steering for Steering Mandate and Steering Opportunity test.

According to Multi-AP Specification v1.0, the client steering control 
message enable efficient steering of STAs between BSSs in a Multi-AP 
network. There are two message types:

1.  steering mandate : 
Request Mode bit set to one indicating a Steering Mandate.
The controller uses the Steering Mandate mechanism to
mandate an agent to attempt steering of one or more associated STAs. 


2. steering opportunity :
Request Mode bit set to zero indicating a Steering Opportunity.
The controller uses the Steering Opportunity mechanism to provide a time
window for an agent to steer one or more associated STAs.
The agent may attempt to steer the STA(s) identified in the request to any
other BSS in the Multi-AP network that the Multi-AP Agent has identified as
suitable target BSS(s) 

#212 #214

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4729,6 +4731,14 @@ bool slave_thread::handle_client_capability_query(Socket *sd, ieee1905_1::CmduMe
+bool slave_thread::handle_client_steering_request(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received CLIENT_STEERING_REQUEST_MESSAGE , mid=" << std::dec << int(mid);
```
I would print the mid in hex format (std::hex). It is easier to read it that way.
(nitpick)

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -902,6 +904,13 @@ bool master_thread::handle_cmdu_1905_channel_preference_report(Socket *sd,
+bool master_thread::handle_cmdu_1905_ack_message(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    auto mid = cmdu_rx.getMessageId();
+    LOG(INFO) << "Received ACK_MESSAGE, mid=" << std::dec << int(mid);
```
std::hex (nitpick)

morantr commented on controller/src/beerocks/master/beerocks_master_mrouter.cpp:
```diff
@@ -55,6 +55,7 @@ bool master_mrouter::init()
         ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_WSC_MESSAGE,
         ieee1905_1::eMessageType::CHANNEL_PREFERENCE_REPORT_MESSAGE,
         ieee1905_1::eMessageType::CHANNEL_SELECTION_RESPONSE_MESSAGE,
+        ieee1905_1::eMessageType::ACK_MESSAGE,
```
Probably it will be best if you add it last, I can see how clang will make it ugly that way...

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4748,6 +4741,27 @@ bool slave_thread::handle_ap_capability_query(Socket *sd, ieee1905_1::CmduMessag
+        LOG(ERROR) << "cmdu creation of type ACK_MESSAGE, has failed";
+        return false;
+    }
+    auto error_code_tlv = cmdu_tx.addClass<wfa_map::tlvErrorCode>();
```
AFAIU, we need to add error_code_tlv only if the station received in the command is not connected to the AP.
I don't think we want to simulate error scenarios.
What actually does missing is sending back CMDU of BTM REPORT (0x9c).

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4759,7 +4760,28 @@ bool slave_thread::handle_client_steering_request(Socket *sd, ieee1905_1::CmduMe
+    }
+
+    auto seering_btm_report_tlv = cmdu_tx.addClass<wfa_map::tlvSteeringBTMReport>();
+    if (!error_code_tlv) {
```
you check the wrong tlv (guess it is a copy-paste error)

CoralMalachi replied:
you are right, thanks!

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4759,7 +4760,28 @@ bool slave_thread::handle_client_steering_request(Socket *sd, ieee1905_1::CmduMe
+    // the BSSID, Operating Class and Channel Number of the identified target BSS
+
+    // build and send Client Steering BTM report message
+    if (!cmdu_tx.create(0, ieee1905_1::eMessageType::CLIENT_STEERING_BTM_REPORT_MESSAGE)) {
```
Oh I see you added the btm response here

morantr commented:
Please do all the changes related to the handling of the steer request and sending the response to a single commit.

BTW, I assigned myself to this flow before the weekend and was surprised to see you have taken it.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -149,7 +149,7 @@ test_init() {
         exit 1
     }
-    AL_MAC=$(docker exec -it gateway ${installdir}/bin/beerocks_cli -c bml_conn_map | grep IRE_BRIDGE | awk '{print $5}' | cut -d ',' -f 1)
+    AL_MAC=$(docker exec -it gateway ${installdir}/bin/beerocks_cli -c bml_conn_map | grep IRE_BRIDGE| head -1 | awk '{print $5}' | cut -d ',' -f 1)
```
I would replace this with an explicit `aa:bb:cc:dd:00:30` (or whatever the correct MAC address is). It may feel a little fragile, but this entire script heavily relies on the exact way that we set up containers that I don't think relying on a specific MAC address is too much to ask.

(nitpick)

arnout commented on tools/docker/tests/test_gw_repeater.sh:
```diff
@@ -60,6 +60,9 @@ main() {
         ${scriptdir}/../run.sh ${VERBOSE_OPT} ${FORCE_OPT} start-agent -d -n ${REPEATER_NAME} -m aa:bb:cc:dd "$@"
+        
+        status "Start Repeater2 (Remote Agent2)"
+        ${scriptdir}/../run.sh ${VERBOSE_OPT} ${FORCE_OPT} start-agent -d -n ${REPEATER2_NAME} -m aa:bb:cc:ee "$@"
```
I would prefer that there would be a command-line option that specifies the number of agents, and that they're started in a loop. The MAC address can be set to `aa:bb:cc:$n$n` where n is the iterator.

tomereli replied:
Agree with Arnout, we can update the `-r|--repeater` argument as follows:
```
while true; do
    case "$1" in
        ...
        -r | --repeater) REPEATER_NAME="$REPEATER_NAME $2"; shift; shift ;;
        ...
    esac
done
```

And then create multiple repeaters by iterating through `REPEATER_NAME` as follows:
```
for repeater in $REPEATER_NAME
do
    # Start repeater with $repeater name
done
```

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -478,6 +480,10 @@ bool slave_thread::handle_cmdu_control_ieee1905_1_message(Socket *sd,
         return handle_channel_selection_request(sd, cmdu_rx);
+    case ieee1905_1::eMessageType::CLIENT_STEERING_REQUEST_MESSAGE:
+        return handle_client_steering_request(sd, cmdu_rx);
+    case ieee1905_1::eMessageType::ACK_MESSAGE:
```
I may get in conflict with @morantr here :-) but I prefer a separate commit for the handling of the ACK message.

However, at least on the longer term, I think it's more appropriate if the ACK messages are handled in the transport. I believe that retransmissions are currently done there as well.

morantr replied:
Where did you get this idea? I have the same opinion as you. :)

CoralMalachi replied:
I'll update my changes soon.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4738,6 +4744,50 @@ bool slave_thread::handle_ap_capability_query(Socket *sd, ieee1905_1::CmduMessag
+        LOG(ERROR) << "cmdu creation of type ACK_MESSAGE, has failed";
+        return false;
+    }
+
```
There should be a TODO here that we have to validate the BTM request before sending the ack.

arnout commented on controller/src/beerocks/master/beerocks_master_mrouter.cpp:
```diff
@@ -55,7 +55,9 @@ bool master_mrouter::init()
         ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_WSC_MESSAGE,
         ieee1905_1::eMessageType::CHANNEL_PREFERENCE_REPORT_MESSAGE,
         ieee1905_1::eMessageType::CHANNEL_SELECTION_RESPONSE_MESSAGE,
+        ieee1905_1::eMessageType::CLIENT_STEERING_BTM_REPORT_MESSAGE,
```
Again, contrary to @morantr, I think the controller and the agent part should be in separate commits.

arnout commented on framework/tlvf/tlvf_conf.yaml:
```diff
@@ -18,6 +19,8 @@ include_yaml_path: {
   "tlvf/wfa_map/tlvRadioOperationRestriction.yaml",
   "tlvf/wfa_map/tlvChannelSelectionResponse.yaml",
   "tlvf/wfa_map/tlvTransmitPowerLimit.yaml",
+  "tlvf/wfa_map/tlvSteeringBTMReport.yaml",
```
This should definitely be a separate commit. Also, the AutoGenerated seems to be missing.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -902,6 +906,32 @@ bool master_thread::handle_cmdu_1905_channel_preference_report(Socket *sd,
+bool master_thread::handle_cmdu_1905_ack_message(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    auto mid = cmdu_rx.getMessageId();
+    LOG(INFO) << "Received ACK_MESSAGE, mid=" << std::hex << int(mid);
```
For debugging it will be useful to also print the error code, if any.

Commit:
agent: Add client steering mandate request handler

According to section 5.8.1 in the EasyMesh Test Plan v1.0 [1],
the MCUT (MultiAP Controller Under Test) sends a client steering request
message to CTT agent 1.

To make sure the steering request message was received by the agent:

1. The agent subscribes to a message of type :
   CLIENT_STEERING_REQUEST_MESSAGE.

2. Add handle_client_steering_request method which print "Received
   CLIENT_STEERING_REQUEST_MESSAGE" when the query is received.
   In addition, create ACK CMDU and BTM report CMDU to send back to the
   controller.

3. If a Multi-AP Agent receives a Client Steering Request message with
   Request Mode bit set to zero indicating a Steering Opportunity to a Multi-AP
   Agent.

   In that case, CTT agent 1 sends a Steering Completed message back to
   the MCUT.Therefore, create a Steering Completed message and send it
   back to the controller.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
controller Add client steering btm report handler

According to section 5.8.1 in the EasyMesh Test Plan v1.0 [1],
when CTT agent 1 receives  a client steering request message, a
client steering BTM report needs to be sent back to the controller.

Therefore, the following changes were made:

1. The controller subscribes to a message of type: Client Steering
   BTM Report.

2. Add handle method which responsible for handling the Client Steering
   BTM Report when receiving at the controller, and print: "Received
   Client Steering BTM Report".
   In addition, create an ACK CMDU to send back to the agent.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: Add ack handler

According to section 5.8.1 in the EasyMesh Test Plan v1.0 [1],
when the MCUT receives a client steering BTM report, an
ACK message needs to be sent back to the agent.

Therefore, the following changes were made:

1. The agent subscribes to a message of type: ACK_MESSAGE.

2. Add handle method which responsible for handling the ACK when
   receiving at the agent, and print: "Received ACK_MESSAGE".

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
controller: Add steering completed message handler

According to section 5.8.1 in the EasyMesh Test Plan v1.0 [1],
when CTT agent 1 receives a client steering request message for Steering
Opportunity, a steering completed message needs to be sent back to the
controller. Therefore, the following changes were made:

1. The controller subscribes to a message of type:
   STEERING_COMPLETED_MESSAGE

2. Add handle method which responsible for handling the Steering
   Completed message when receiving at the controller, and print:
   "Received STEERING_COMPLETED_MESSAGE".
   In addition, create an ACK CMDU to send back to the agent.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
controller,agent: Remove handle cmdu message

Remove printing of handle_cmdu since it is not necessary anymore now when
we have a handle method for the ACK message. Therefore, in test_flows.sh
search for ACK_MESSAGE.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tools: docker: Add client steering flow test

Add a test that triggers the sending of a Client Steering Request
message using the WFA_CA format utilized by the WFA_UCC testbed.

The test flow sends the command and making sure that the client
steering request was received by the repeater and the entire flow
continued as expected.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4741,6 +4744,42 @@ bool slave_thread::handle_ap_capability_query(Socket *sd, ieee1905_1::CmduMessag
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received CLIENT_STEERING_REQUEST_MESSAGE , mid=" << std::hex << int(mid);
+    //TODO: validate the BTM request before sending the ack.
```
I think the minimum validation we should to is to check the BSSID. Indeed, both slaves are handling this request, so both of them will send an Ack and a BTM Report. This is clearly wrong.

However, it doesn't matter for certification, and this will anyway be rewritten extensively when we complete the agent implementation, so (nitpick).

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -222,13 +222,17 @@ bool master_thread::handle_cmdu_1905_1_message(Socket *sd, ieee1905_1::CmduMessa
     }
 
-    LOG(WARNING) << "Unknown 1905 message received. Ignoring";
+    LOG(WARNING) << "Unknown 1905 message received. Ignoring, Message mid=" << std::hex
```
Good change, but probably should have been a separate commit. (nitpick)

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4744,6 +4746,15 @@ bool slave_thread::handle_ap_capability_query(Socket *sd, ieee1905_1::CmduMessag
     return true;
 }
 
+bool slave_thread::handle_ack_request(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
```
I'm ambivalent about this one... As per #273 there are two options for handling acks: either in the transport, or in each component. However, if acks are handled in the component, they will need to be handled both in the backhaul manager (for messages that are emitted from there) and in the slaves - and the backhaul manager should only forward the acks that do not correspond to a message it sent itself.

I guess for now, however, it's OK to handle the acks in both slaves.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4759,6 +4759,10 @@ bool slave_thread::handle_client_steering_request(Socket *sd, ieee1905_1::CmduMe
     const auto mid = cmdu_rx.getMessageId();
     LOG(DEBUG) << "Received CLIENT_STEERING_REQUEST_MESSAGE , mid=" << std::hex << int(mid);
+    auto rx_buffer        = cmdu_rx.getMessageBuff();
+    uint16_t request_mode = (uint16_t)(rx_buffer[17]);
```
I really don't like this... The TLV has been defined already, so why not do the proper addClass and get the request_mode?

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4776,18 +4780,33 @@ bool slave_thread::handle_client_steering_request(Socket *sd, ieee1905_1::CmduMe
+        LOG(DEBUG) << "sending CLIENT_STEERING_BTM_REPORT_MESSAGE back to controller";
+    } else {
+        LOG(DEBUG) << "Request Mode bit is not set - Steering Opportunity";
+        // build and send steering complited message
```
complited -> completed

arnout reviewed (changes requested):
There's a typo in the "remove handle_cmdu" commit: neseccary -> necessary

I do want the parsing of the steering request TLV to be fixed before merging. The rest are nitpicks or general remarks.

CoralMalachi commented:
@tomereli @arnout 
Do you want me to do rebase and autosquash all the fixup commits?

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -961,10 +963,26 @@ bool master_thread::handle_cmdu_1905_channel_preference_report(Socket *sd,
+        LOG(ERROR) << "cmdu creation of type ACK_MESSAGE, has failed";
+        return false;
+    }
+    LOG(DEBUG) << "sending ACK message back to agent";
```
Add MID to the log

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4759,8 +4760,14 @@ bool slave_thread::handle_client_steering_request(Socket *sd, ieee1905_1::CmduMe
     LOG(DEBUG) << "Received CLIENT_STEERING_REQUEST_MESSAGE , mid=" << std::hex << int(mid);
-    auto rx_buffer        = cmdu_rx.getMessageBuff();
-    uint16_t request_mode = (uint16_t)(rx_buffer[17]);
+    auto steering_btm_report_tlv = cmdu_rx.addClass<wfa_map::tlvSteeringRequest>();
```
It's a steering request, not a BTM report.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4759,8 +4760,14 @@ bool slave_thread::handle_client_steering_request(Socket *sd, ieee1905_1::CmduMe
+        return false;
+    }
+
+    // auto x = steering_btm_report_tlv->request_flags()
```
What does this comment mean?

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4741,6 +4745,76 @@ bool slave_thread::handle_ap_capability_query(Socket *sd, ieee1905_1::CmduMessag
+    // the BSSID, Operating Class and Channel Number of the identified target BSS
+
+    //check if the Request Mode bit is set
+    if ((request_mode >> (7)) & 1) {
```
Is this correct? I thought that request_mode was a one-bit value already. IOW, I'd expect
```suggestion
    if (request_mode) {
```

CoralMalachi replied:
Yes, you are right, that wad before I used the addClass, I'll change it now.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -222,13 +220,21 @@ bool master_thread::handle_cmdu_1905_1_message(Socket *sd, ieee1905_1::CmduMessa
 
-    LOG(WARNING) << "Unknown 1905 message received. Ignoring";
+    LOG(WARNING) << "Unknown 1905 message received. Ignoring, Message mid=" << std::hex
+                 << int(cmdu_rx.getMessageType());
```
I didn't notice before: you're not printing the MID here, but the type. You probably want both.

arnout reviewed (changes requested):
Please rebase -i on master for a final review (and fixup the small remaining issues in the process).

Make sure the tlv_conf.yaml + AutoGenerated update are in a separate commit.

Commit:
tlvf: tlvSteeringRequest auto-generated files update

Preparative commit.
Add "tlvf/wfa_map/tlvSteeringRequest.yaml" to tlv_conf.yaml file
and add its auto generated files as well.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

morantr committed:
tlvf: tlvSteeringBTMReport auto-generated files update

Preparative commit.
Add "tlvf/wfa_map/tlvSteeringBTMReport.yaml" to tlv_conf.yaml file
and add its auto generated files as well.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
